snow/networking/worker/pool.go

```diff
@@ -3,60 +3,118 @@
 
 package worker
 
-import "sync"
+import (
+	"errors"
+	"sync"
+)
 
-var _ Pool = (*pool)(nil)
+var (
+	_ Pool = (*pool)(nil)
+
+	ErrNoWorkers = errors.New("attempting to create worker pool with less than 1 worker")
+)
 
 type Request func()
 
 type Pool interface {
+	// Starts the worker pool.
+	//
+	// THis method should be called before Send and Shutdown.
+	Start()
+
 	// Send the request to the worker pool.
 	//
-	// Send should never be called after [Shutdown] is called.
+	// Send can be safely called before [Start] and it'll be no-op.
+	// Send can be safely called after [Shutdown] and it'll be no-op.
 	Send(Request)
 
 	// Shutdown the worker pool.
 	//
 	// This method will block until all workers have finished their current
 	// tasks.
 	//
-	// It is safe to call shutdown multiple times.
+	// Shutdown can be safely called before [Start] and it'll be no-op.
+	// Shutdown can be safely called multiple times.
 	Shutdown()
 }
 
 type pool struct {
-	requests chan Request
+	workersCount int
+	requests     chan Request
 
+	startOnce    sync.Once
 	shutdownOnce sync.Once
-	shutdownWG   sync.WaitGroup
+
+	// [shutdownWG] makes sure all workers have stopped before Shutdown returns
+	shutdownWG sync.WaitGroup
+
+	// closing [quit] tells the workers to stop working
+	quit chan struct{}
 }
 
-func NewPool(size int) Pool {
-	p := &pool{
-		requests: make(chan Request),
+func NewPool(workersCount int) (Pool, error) {
+	if workersCount <= 0 {
+		return nil, ErrNoWorkers
 	}
-	p.shutdownWG.Add(size)
-	for w := 0; w < size; w++ {
-		go p.runWorker()
+
+	p := &pool{
+		workersCount: workersCount,
 	}
-	return p
+
+	// Note: we instantiate requests and quit channels
+	// only when start is called.
+	return p, nil
+}
+
+func (p *pool) Start() {
```

Comment:
1. Do we expect `Start` to be called multiple times? I see we're using `sync.Once` to guard against it.
2. We can also get rid of this edge-case by leaving the goroutine initialization in `NewPool`. Instead of having a `Shutdown` code we can pass in a channel or a context into `NewPool` so the caller can shutdown the pool by calling context cancel or by closing the shutdown channel. This will also simplify the edge-cases around `Send` since it'll only be possible to call `Send` on an already running pool.

Reply:
Hmm actually found that `errgroup.SetLimit` pretty much implements a goroutine pool for us. I think if we just wrap an errgroup with a shutdown flag we'll pretty much have our worker pool. ref: https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit